src/browser/connection/index.js

```diff
@@ -0,0 +1,246 @@
+        this.heartbeatTimeout = null;
+
+        this.url           = `${gateway.domain}/browser/connect/${this.id}`;
+        this.heartbeatUrl  = `${gateway.domain}/browser/heartbeat/${this.id}`;
+        this.idleUrl       = `${gateway.domain}/browser/idle/${this.id}`;
+        this.statusUrl     = `${gateway.domain}/browser/status/${this.id}`;
+        this.initScriptUrl = `${gateway.domain}/browser/init-script/${this.id}`;
+
+        this.on('error', () => {
+            this._forceIdle();
+            this.close();
+        });
+
+        connections[this.id] = this;
+
+        this.browserConnectionGateway.startServingConnection(this);
+
+        this._runBrowser();
+    }
+
+    static _generateId () {
+        return shortId.generate();
+    }
+
+    _runBrowser () {
+        // NOTE: Give caller time to assign event listeners
+        process.nextTick(async () => {
+            try {
+                await this.provider.openBrowser(this.id, this.browserInfo.browserName, this.url);
+
+                if (!this.ready)
+                    await promisifyEvent(this, 'ready');
+
+                this.opened = true;
+                this.emit('opened');
+            }
+            catch (err) {
+                this.emit('error', new GeneralError(
+                    MESSAGE.browserOpeningFailed,
+                    this.browserInfo.providerName + ':' + this.browserInfo.browserName,
+                    err.message
+                ));
+            }
+        });
+    }
+
+    async _closeBrowser () {
+        if (!this.idle)
+            await promisifyEvent(this, 'idle');
+
+        try {
+            await this.provider.closeBrowser(this.id, { title: this.idleUrl, url: this.idleUrl });
+        }
+        catch (err) {
+            // NOTE: A warning would be a really nice here, but it can't be done while log is stored in a task.
+        }
+    }
+
+    _forceIdle () {
+        if (!this.idle) {
+            this.switchingToIdle = false;
+            this.idle            = true;
+            this.emit('idle');
+        }
+    }
+
+    _waitForHeartbeat () {
+        this.heartbeatTimeout = setTimeout(() => {
+            this.emit('error', new GeneralError(MESSAGE.browserDisconnected, this.userAgent));
+        }, this.HEARTBEAT_TIMEOUT);
+    }
+
+    _popNextTestRunUrl () {
+        while (this.hasQueuedJobs && !this.currentJob.hasQueuedTestRuns)
+            this.jobQueue.shift();
+
+        return this.hasQueuedJobs ? this.currentJob.popNextTestRunUrl() : null;
+    }
+
+    static getById (id) {
+        return connections[id] || null;
+    }
+
+    addWarning (...args) {
+        if (!this.currentJob)
+            return;
+
+        this.currentJob.warningLog.addWarning(...args);
+    }
+
+    setUserAgentCustomPart (str) {
+        this.browserInfo.userAgentCustomPart = str;
+    }
+
+    get userAgent () {
+        var userAgent = this.browserInfo.userAgent;
+
+        if (this.browserInfo.userAgentCustomPart)
```

Comment:
Maybe rename it to `userAgentAdditionalInfo` or even better `userAgentProviderMetaInfo`